[PROPOSAL] Unable to extract subtitles from DVB video samples when build using cmake by passing -DWITH_FFMPEG=ON flag

Please prefix your issue with one of the following: [BUG], [PROPOSAL], [QUESTION].

CCExtractor version (using the --version parameter preferably) : **X.X**

**In raising this issue, I confirm the following (please check boxes, eg [X] - and delete unchecked ones):**

- [x ] I have read and understood the [contributors guide](https://github.com/CCExtractor/ccextractor/blob/master/.github/CONTRIBUTING.md).
- [x ] I have checked that the bug-fix I am reporting can be replicated, or that the feature I am suggesting isn't already present.
- [x ] I have checked that the issue I'm posting isn't already reported.
- [x ] I have checked that the issue I'm porting isn't already solved and no duplicates exist in [closed issues](https://github.com/CCExtractor/ccextractor/issues?q=is%3Aissue+is%3Aclosed) and in [opened issues](https://github.com/CCExtractor/ccextractor/issues)
- [x ] I have checked the pull requests tab for existing solutions/implementations to my issue/suggestion.
- [x ] I have used the latest available version of CCExtractor to verify this issue exists.

**My familiarity with the project is as follows (check one, eg [X] - and delete unchecked ones):**

- [ ] I have never used CCExtractor.
- [ ] I have used CCExtractor just a couple of times.
- [ ] I absolutely love CCExtractor, but have not contributed previously.
- [x ] I am an active contributor to CCExtractor.

**Necessary information**
- Is this a regression (did it work before)? [x ] NO | [ ] YES - *please specify the last known working version*
- What platform did you use? [ ] Windows - [x ] Linux - [ ] Mac
- What were the used arguments? `-autoprogram`

**Video links (replace text below with your links) **
https://drive.google.com/open?id=0B_61ywKPmI0TdTdSTi1iN251WHM

**Additional information**

Unable to extract subtitles from DVB video samples.
![screenshot from 2018-05-29 15-32-00](https://user-images.githubusercontent.com/31391329/40652684-918fceee-6356-11e8-9c63-5ac39cf71c64.png)

This occurs if I build ccextractor using cmake by passing the -DWITH_FFMPEG=ON flag
Exact build steps : 
![screenshot from 2018-05-29 15-34-05](https://user-images.githubusercontent.com/31391329/40652907-43752be0-6357-11e8-9dc1-de8a91e48665.png)

If I build without the -DWITH_FFMPEG=ON flag, ccextractor is able to extract the subtitles
![screenshot from 2018-05-29 15-46-47](https://user-images.githubusercontent.com/31391329/40653035-8de80760-6357-11e8-850b-b87a67e3bbff.png)


PS: Make sure you set an alert in GitHub so you get notifications about your ticket. We may need to ask questions and we do everything inside GitHub's system.

Name of file: 2016-12-15-BBC4.ts  (leaving in a comment so it's findable)

I don't get those errors but I also don't get any output (no captions found). There's captions for sure (VLC plays them) so this is still a valid bug

```
✦ ❯ ./build -debug -hardsubx && ./ccextractor ~/Downloads/2016-12-15-BBC4.ts
Running pre-build script...
Obtaining Git commit
Storing variables in file
Commit: 9b2215d9c2ed48cff1cd9f1fa73af6e1bc88621f
Date: 2023-03-22
Stored all in compile_info_real.h
Done.
Trying to compile...
Checking for cargo...
rustc >= MSRV(1.54.0)
Building rust files...
    Finished dev [unoptimized + debuginfo] target(s) in 0.02s
Building ccextractor
../src/lib_ccx/telxcc.c: In function 'telxcc_dump_prev_page':
../src/lib_ccx/telxcc.c:519:28: warning: '%.3u' directive output may be truncated writing between 3 and 5 bytes into a region of size 4 [-Wformat-truncation=]
  519 |         snprintf(info, 4, "%.3u", bcd_page_to_int(tlt_config.page));
      |                            ^~~~
../src/lib_ccx/telxcc.c:519:27: note: directive argument in the range [0, 65535]
  519 |         snprintf(info, 4, "%.3u", bcd_page_to_int(tlt_config.page));
      |                           ^~~~~~
../src/lib_ccx/telxcc.c:519:9: note: 'snprintf' output between 4 and 6 bytes into a destination of size 4
  519 |         snprintf(info, 4, "%.3u", bcd_page_to_int(tlt_config.page));
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from ../src/lib_ccx/matroska.c:3:
../src/lib_ccx/matroska.c: In function 'generate_timestamp_ass_ssa':
../src/lib_ccx/matroska.h:23:13: warning: '%02ld' directive writing 2 bytes into a region of size between 1 and 13 [-Wformat-overflow=]
   23 | #define LLD "%ld"
      |             ^~~~~
../src/lib_ccx/matroska.c:256:22: note: in expansion of macro 'LLD'
  256 |         sprintf(buf, LLD ":%02" LLD_M ":%02" LLD_M ".%02" LLD_M, hours, minutes, seconds, millis);
      |                      ^~~
../src/lib_ccx/matroska.c:256:28: note: format string is defined here
  256 |         sprintf(buf, LLD ":%02" LLD_M ":%02" LLD_M ".%02" LLD_M, hours, minutes, seconds, millis);
../src/lib_ccx/matroska.h:23:13: note: directive argument in the range [0, 59]
   23 | #define LLD "%ld"
      |             ^~~~~
../src/lib_ccx/matroska.c:256:22: note: in expansion of macro 'LLD'
  256 |         sprintf(buf, LLD ":%02" LLD_M ":%02" LLD_M ".%02" LLD_M, hours, minutes, seconds, millis);
      |                      ^~~
../src/lib_ccx/matroska.h:23:13: note: directive argument in the range [0, 59]
   23 | #define LLD "%ld"
      |             ^~~~~
../src/lib_ccx/matroska.c:256:22: note: in expansion of macro 'LLD'
  256 |         sprintf(buf, LLD ":%02" LLD_M ":%02" LLD_M ".%02" LLD_M, hours, minutes, seconds, millis);
      |                      ^~~
../src/lib_ccx/matroska.h:23:13: note: directive argument in the range [0, 99]
   23 | #define LLD "%ld"
      |             ^~~~~
../src/lib_ccx/matroska.c:256:22: note: in expansion of macro 'LLD'
  256 |         sprintf(buf, LLD ":%02" LLD_M ":%02" LLD_M ".%02" LLD_M, hours, minutes, seconds, millis);
      |                      ^~~
../src/lib_ccx/matroska.c:256:9: note: 'sprintf' output between 11 and 23 bytes into a destination of size 15
  256 |         sprintf(buf, LLD ":%02" LLD_M ":%02" LLD_M ".%02" LLD_M, hours, minutes, seconds, millis);
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../src/lib_ccx/matroska.c: In function 'save_sub_track':
../src/lib_ccx/matroska.c:1199:42: warning: '%d' directive writing between 1 and 10 bytes into a region of size 9 [-Wformat-overflow=]
 1199 |                         sprintf(number, "%d", i + 1);
      |                                          ^~
../src/lib_ccx/matroska.c:1199:41: note: directive argument in the range [1, 2147483647]
 1199 |                         sprintf(number, "%d", i + 1);
      |                                         ^~~~
../src/lib_ccx/matroska.c:1199:25: note: 'sprintf' output between 2 and 11 bytes into a destination of size 9
 1199 |                         sprintf(number, "%d", i + 1);
      |                         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
../src/lib_ccx/utility.c: In function 'change_filename':
../src/lib_ccx/utility.c:457:50: warning: '%d' directive writing between 1 and 10 bytes into a region of size 5 [-Wformat-overflow=]
  457 |                         sprintf(new_extension, ".%d", iter);
      |                                                  ^~
../src/lib_ccx/utility.c:457:48: note: directive argument in the range [1, 2147483647]
  457 |                         sprintf(new_extension, ".%d", iter);
      |                                                ^~~~~
../src/lib_ccx/utility.c:457:25: note: 'sprintf' output between 3 and 12 bytes into a destination of size 6
  457 |                         sprintf(new_extension, ".%d", iter);
      |                         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../src/lib_ccx/ts_tables_epg.c: In function 'EPG_DVB_calc_start_time':
../src/lib_ccx/ts_tables_epg.c:85:57: warning: '%02ld' directive writing between 2 and 20 bytes into a region of size between 1 and 19 [-Wformat-overflow=]
   85 |                 sprintf(event->start_time_string, "%02ld%02ld%02ld%06" PRIu64 "+0000", y, m, d, time & 0xffffff);
      |                                                         ^~~~~
../src/lib_ccx/ts_tables_epg.c:85:51: note: directive argument in the range [-9223372036854775808, 9223372036854775806]
   85 |                 sprintf(event->start_time_string, "%02ld%02ld%02ld%06" PRIu64 "+0000", y, m, d, time & 0xffffff);
      |                                                   ^~~~~~~~~~~~~~~~~~~~
../src/lib_ccx/ts_tables_epg.c:85:51: note: using the range [-9223372036854775808, 9223372036854775807] for directive argument
../src/lib_ccx/ts_tables_epg.c:85:51: note: directive argument in the range [0, 16777215]
../src/lib_ccx/ts_tables_epg.c:85:17: note: 'sprintf' output between 18 and 74 bytes into a destination of size 21
   85 |                 sprintf(event->start_time_string, "%02ld%02ld%02ld%06" PRIu64 "+0000", y, m, d, time & 0xffffff);
      |                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../src/lib_ccx/ts_tables_epg.c: In function 'EPG_DVB_calc_end_time':
../src/lib_ccx/ts_tables_epg.c:116:54: warning: '%02d' directive writing between 2 and 11 bytes into a region of size between 10 and 19 [-Wformat-overflow=]
  116 |                 sprintf(event->end_time_string, "%02d%02d%02d%02d%02d%02d +0000", timeinfo.tm_year + 1900, timeinfo.tm_mon + 1, timeinfo.tm_mday, timeinfo.tm_hour, timeinfo.tm_min, timeinfo.tm_sec);
      |                                                      ^~~~
../src/lib_ccx/ts_tables_epg.c:116:49: note: directive argument in the range [-2147483647, 2147483647]
  116 |                 sprintf(event->end_time_string, "%02d%02d%02d%02d%02d%02d +0000", timeinfo.tm_year + 1900, timeinfo.tm_mon + 1, timeinfo.tm_mday, timeinfo.tm_hour, timeinfo.tm_min, timeinfo.tm_sec);
      |                                                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../src/lib_ccx/ts_tables_epg.c:116:17: note: 'sprintf' output between 19 and 73 bytes into a destination of size 21
  116 |                 sprintf(event->end_time_string, "%02d%02d%02d%02d%02d%02d +0000", timeinfo.tm_year + 1900, timeinfo.tm_mon + 1, timeinfo.tm_mday, timeinfo.tm_hour, timeinfo.tm_min, timeinfo.tm_sec);
      |                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Compilation successful, compiler message shown in previous lines
CCExtractor 0.94, Carlos Fernandez Sanz, Volker Quetschke.
Teletext portions taken from Petr Kutalek's telxcc
--------------------------------------------------------------------------
Input: /home/cfsmp3/Downloads/2016-12-15-BBC4.ts
[Extract: 1] [Stream mode: Autodetect]
[Program : Auto ] [Hauppage mode: No] [Use MythTV code: Auto]
[CEA-708: 63 decoders active]
[CEA-708: using charset "none" for all services]
[Timing mode: Auto] [Debug: No] [Buffer input: No]
[Use pic_order_cnt_lsb for H.264: No] [Print CC decoder traces: No]
[Target format: .srt] [Encoding: UTF-8] [Delay: 0] [Trim lines: No
```

--- a/src/lib_ccx/ocr.c
+++ b/src/lib_ccx/ocr.c
@@ -696,23 +696,15 @@ char *ocr_bitmap(void *arg, png_color *palette, png_byte *alpha, unsigned char *
 	return text_out;
 }
 
-void erode(png_color *palette, png_byte *alpha, uint8_t *bitmap, int w, int h, int nb_color)
+void erode(png_color *palette, png_byte *alpha, uint8_t *bitmap, int w, int h, int nb_color, int background_index)
 {
-	int background_index;
-	for (background_index = 0; background_index < nb_color; background_index++)
-	{
-		if (alpha[background_index])
-		{
-			break;
-		}
-	}
 	// we will use a 2*2 kernel for the erosion
 	for (int row = 0; row < h - 1; row++)
 	{
 		for (int col = 0; col < w - 1; col++)
 		{
-			if (alpha[bitmap[row * w + col]] || alpha[bitmap[(row + 1) * w + col]] ||
-			    alpha[bitmap[row * w + (col + 1)]] || alpha[bitmap[(row + 1) * w + (col + 1)]])
+			if (bitmap[row * w + col] == background_index || bitmap[(row + 1) * w + col] == background_index ||
+			    bitmap[row * w + (col + 1)] == background_index || bitmap[(row + 1) * w + (col + 1)] == background_index)
 			{
 				bitmap[row * w + col] = background_index;
 			}
@@ -720,23 +712,15 @@ void erode(png_color *palette, png_byte *alpha, uint8_t *bitmap, int w, int h, i
 	}
 }
 
-void dilate(png_color *palette, png_byte *alpha, uint8_t *bitmap, int w, int h, int nb_color)
+void dilate(png_color *palette, png_byte *alpha, uint8_t *bitmap, int w, int h, int nb_color, int foreground_index)
 {
-	int foreground_index;
-	for (foreground_index = 0; foreground_index < nb_color; foreground_index++)
-	{
-		if (!alpha[foreground_index])
-		{
-			break;
-		}
-	}
 	// we will use a 2*2 kernel for the erosion
 	for (int row = 0; row < h - 1; row++)
 	{
 		for (int col = 0; col < w - 1; col++)
 		{
-			if (!(alpha[bitmap[row * w + col]] && alpha[bitmap[(row + 1) * w + col]] &&
-			      alpha[bitmap[row * w + (col + 1)]] && alpha[bitmap[(row + 1) * w + (col + 1)]]))
+			if ((bitmap[row * w + col] == foreground_index && bitmap[(row + 1) * w + col] == foreground_index &&
+			     bitmap[row * w + (col + 1)] == foreground_index && bitmap[(row + 1) * w + (col + 1)] == foreground_index))
 			{
 				bitmap[row * w + col] = foreground_index;
 			}
@@ -767,6 +751,7 @@ static int quantize_map(png_byte *alpha, png_color *palette,
 	 */
 	uint32_t *mcit = NULL;
 	struct transIntensity ti = {alpha, palette};
+	int text_color, text_bg_color;
 
 	int ret = 0;
 
@@ -833,6 +818,14 @@ static int quantize_map(png_byte *alpha, png_color *palette,
 				max_ind = j;
 			}
 		}
+
+		// Assume second most frequent color to be text background (first is alpha channel)
+		if (i == 1)
+			text_bg_color = iot[max_ind];
+		// Assume third most frequent color to be text color
+		if (i == 2)
+			text_color = iot[max_ind];
+
 		for (j = i; j > 0 && max_ind < mcit[j - 1]; j--)
 		{
 			mcit[j] = mcit[j - 1];
@@ -876,8 +869,8 @@ static int quantize_map(png_byte *alpha, png_color *palette,
 			palette[iot[i]].green = palette[index].green;
 		}
 	}
-	erode(palette, alpha, bitmap, w, h, nb_color);
-	dilate(palette, alpha, bitmap, w, h, nb_color);
+	erode(palette, alpha, bitmap, w, h, nb_color, text_bg_color);
+	dilate(palette, alpha, bitmap, w, h, nb_color, text_color);
 #ifdef OCR_DEBUG
 	ccx_common_logging.log_ftn("Colors present in quantized Image\n");
 	for (int i = 0; i < nb_color; i++)

--- a/src/lib_ccx/ocr.c
+++ b/src/lib_ccx/ocr.c
@@ -9,7 +9,6 @@
 #include "ccx_encoders_helpers.h"
 #include "ccx_encoders_spupng.h"
 #include "ocr.h"
-#undef OCR_DEBUG
 
 struct ocrCtx
 {
@@ -686,7 +685,6 @@ char *ocr_bitmap(void *arg, png_color *palette, png_byte *alpha, unsigned char *
 		TessResultIteratorDelete(ri);
 	}
 	// End Color Detection
-	freep(&text_out);
 	boxDestroy(&crop_points);
 
 	pixDestroy(&pix);
@@ -1062,7 +1060,13 @@ char *paraof_ocrtext(struct cc_subtitle *sub, struct encoder_ctx *context)
 			len += strlen(rect->ocr_text);
 	}
 	if (len <= 0)
+	{
+		for (i = 0, rect = sub->data; i < sub->nb_data; i++, rect++)
+		{
+			freep(&rect->ocr_text);
+		}
 		return NULL;
+	}
 	else
 	{
 		str = malloc(len + 1 + 10); // Extra space for possible trailing '/n's at the end of tesseract UTF8 text
@@ -1076,7 +1080,7 @@ char *paraof_ocrtext(struct cc_subtitle *sub, struct encoder_ctx *context)
 		if (!rect->ocr_text)
 			continue;
 		add_ocrtext2str(str, rect->ocr_text, context->encoded_crlf, context->encoded_crlf_length);
-		free(rect->ocr_text);
+		freep(&rect->ocr_text);
 	}
 	return str;
 }

